src/pykeen/triples/instances.py

```diff
@@ -64,20 +52,50 @@ def __getitem__(self, item):  # noqa: D105
 class LCWAInstances(Instances):
     """Triples and mappings to their indices for LCWA."""
 
-    labels: np.ndarray
+    #: The unique pairs
+    pairs: np.ndarray
+
+    #: The compressed triples in CSR format
+    compressed: scipy.sparse.csr_matrix
+
+    @classmethod
+    def from_triples(cls, mapped_triples: MappedTriples, num_entities: int) -> "LCWAInstances":
+        """
+        Create LCWA instances from triples.
+
+        :param mapped_triples: shape: (num_triples, 3)
+            The ID-based triples.
+        :param num_entities:
+            The number of entities.
+
+        :return:
+            The instances.
+        """
+        mapped_triples = mapped_triples.numpy()
+        unique_hr, pair_idx_to_triple_idx = np.unique(mapped_triples[:, :2], return_inverse=True, axis=0)
+        num_pairs = unique_hr.shape[0]
+        tails = mapped_triples[:, 2]
+        compressed = scipy.sparse.coo_matrix(
+            (np.ones(mapped_triples.shape[0], dtype=np.float32), (pair_idx_to_triple_idx, tails)),
+            shape=(num_pairs, num_entities)
+        )
+        # convert to csr for fast row slicing
+        compressed = compressed.tocsr()
+        return LCWAInstances(pairs=unique_hr, compressed=compressed)
+
+    def __len__(self) -> int:  # noqa: D105
+        return self.pairs.shape[0]
 
     def __getitem__(self, item):  # noqa: D105
-        # Create dense target
-        batch_labels_full = torch.zeros(self.num_entities)
-        batch_labels_full[self.labels[item]] = 1
-        return self.mapped_triples[item], batch_labels_full
+        return self.pairs[item], np.asarray(self.compressed[item, :].todense())[0, :]
 
 
 @fix_dataclass_init_docs
 @dataclass
 class MultimodalInstances(Instances):
     """Triples and mappings to their indices as well as multimodal data."""
 
+    #: TODO: do we need these?
```

Comment:
it remains a great mystery since we don't have any unit tests running for the literal models

Reply:
Yes, being required to apply any changes also to the literal models without having any way to validate them is also a problem for me :wink: I already thought about suggesting to remove them until they come back with a justification (=test / training script / ...).